[CanonicalizeOSSALifetime] Extend Onone lifetimes.

nate-chandler opened the pull request:
To improve the debugging experience of values whose lifetimes are canonicalized without compromising the semantics expressed in the source language, when canonicalizing OSSA lifetimes at Onone, lengthen lifetimes as much as possible without incurring copies that would be eliminated at O.

rdar://99618502

nate-chandler committed:
[NFC] Removed unused field.

nate-chandler commented:
@swift-ci please smoke test

nate-chandler committed:
[CanonicalizeOSSALifetime] Fixed caching bug.

Previously, the destroys set (now set vector) wasn't ever being cleared.
The result was that users could get overly pessimistic behavior if they
had previously used the utility with a destroy that came after the
destroys relevant for its current run. Here, it is cleared when the
utility is initialized with a new def.  Addresses a TODO in the
copy_propagation test.

nate-chandler commented:
@swift-ci please test

atrick commented on lib/SILOptimizer/Utils/CanonicalizeOSSALifetime.cpp:
```diff
@@ -611,6 +614,121 @@ void CanonicalizeOSSALifetime::insertDestroysOnBoundary(
+    // Walk backwards from consuming blocks.
+    auto *currentDefBlock = currentDef->getParentBlock();
+    while (auto *block = worklist.pop()) {
+      originalLiveBlocks.insert(block);
```
So `liveness.liveBlocks` is already a potentially large subset of `originalLiveBlocks`. computeCanonicalLiveness already did that work for you. A block only needs to be pushed onto the "original live" worklist if it's not originally in liveBlocks. And there's no need to check for the def block. You might need a wrapper to check for membership in both places... but it seems worthwhile.

nate-chandler replied:
That's great, thanks!  Just adding the blocks discovered by liveness to the set here--that way we only iterate the array once.  (Also, more readable than checking membership in two places.)

atrick commented on lib/SILOptimizer/Utils/CanonicalizeOSSALifetime.cpp:
```diff
@@ -611,6 +614,121 @@ void CanonicalizeOSSALifetime::insertDestroysOnBoundary(
+
+  // Second, collect the blocks which occur after a _final_ consuming use.
+  BasicBlockSetVector finalConsumingBlocks(currentDef->getFunction());
+  BasicBlockSetVector consumedBlocks(currentDef->getFunction());
```
It was hard for me to understand the difference between finalConsumingBlocks and consumedBlocks. It turns out that the former is a superset of the later. But the subsequent code treats them like a single set. We have two queries:

`!consumed && !finalConsuming)` and `consumed || finalConsuming`

Both of those are disjunctions. So logically, the union of the two sets would be equivalent. Since one is a superset of the other, you can just delete the smaller set.

You still need a consumedBlocks *vector* for iteration. What you actually want is:

`BasicBlockSet consumedAtExitBlocks`
`SmallVector<SILBasicBlock *> consumedAtEntryBlocks`

That would also makes the code easier to understand.

nate-chandler replied:
Oops, the intention here was to have finalConsumingBlocks just be the blocks that contain the final non-destroy consumes.  So there wouldn't be a subset/superset relationship.  But the line to insert the blocks was in the wrong loop.

Anyways, I agree with your suggestion.  Changed to that.

nate-chandler commented:
@swift-ci please test source compatibility

atrick commented on lib/SILOptimizer/Utils/CanonicalizeOSSALifetime.cpp:
```diff
@@ -611,6 +614,121 @@ void CanonicalizeOSSALifetime::insertDestroysOnBoundary(
+        } else {
+          for (auto *grandPredecessor : predecessor->getPredecessorBlocks()) {
+            liveness.updateForUse(grandPredecessor->getTerminator(),
+                                  /*lifetimeEnding*/ false);
```
Can you try out a test case where `predecessor` is itself a control-flow merge? We currently assume that a phi operand (branch instruction) can never be the last use unless it is lifetime-ending (This needs to be more clearly commented and asserted somewhere in PrunedLiveness). Breaking this assumption probably means we end up visiting the same insertion point multiple times.

One option is to simply insert the destroy_value at the correct place here and make it the live use.

Another option is to add a special case to visitInsertionPoints to handle nonconsuming-last-use-phis with extra book-keeping.

nate-chandler replied:
Oops, it looks like all the tests I added didn't make it into the PR.  Fixed that (copy_propagation_onone).

nate-chandler replied:
Added.  It exposed an issue when attempting to extend a boundary along multiple cfg edges with the same destination.

nate-chandler commented:
@swift-ci please benchmark

atrick commented on lib/SILOptimizer/Utils/CanonicalizeOSSALifetime.cpp:
```diff
@@ -781,6 +899,11 @@ bool CanonicalizeOSSALifetime::canonicalizeValueLifetime(SILValue def) {
+  if (ononeMode) {
+    extendUnconsumedLiveness(boundary);
+    boundary.clear();
+    findExtendedBoundary(boundary);
```
It looks like `findExtendedBoundary` supposed to initialize a new boundary? If so, please assert that the boundary is empty inside that utility.

nate-chandler replied:
That's right.  An assertion would be nice, I agree.  Added.

atrick reviewed (changes requested):
This is awesome. It's such a relief to see all the nasty stuff that accumulated because of poison values and debug support stripped away and have everything composed on top of our incremental liveness utility. This code is really clear, solid, and maintainable now

nate-chandler commented on include/swift/SILOptimizer/Utils/CanonicalizeOSSALifetime.h:
```diff
@@ -205,6 +205,9 @@ class CanonicalizeOSSALifetime {
   bool pruneDebugMode;
 
+  /// If true, lifetimes will not be shortened.
+  bool ononeMode;
```
Changed to `maximizeLifetime`.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

atrick reviewed (approved):
Beautiful!

The messiness of reusing existing destroys based on some heuristic is perfectly encapsulated exactly where it should be. Not hidden behind any clever shenanigans.

nate-chandler committed:
[CanonicalizeOSSALifetime] Minor cleanup.

nate-chandler committed:
[Gardening] Tweaked comments.

nate-chandler committed:
[CanonicalizeOSSALifetime] Handle live phis.

It is possible for phis to be marked live.  With guaranteed phis, they
will be the last uses and be non-consuming.  In this case, the
merge block will have multiple predecessors whose terminators are on the
boundary.  When inserting destroys, track whether a merge point has been
visited previously.

To facilitate this, restructure the boundary extension and destroy
insertion code.

Previously, the extender was building up a list of places at which to
insert destroys.  In particular it was using the "boundary edge"
collection for all blocks at the beginning of which a destroy should be
created.  In particular, it would add merge blocks. Such blocks are not
boundary blocks.

Here, the extender produces a PrunedLivenessBoundary which doesn't
violate that invariant.

This required some changes to the destroy insertion code to find where
to insert destroys.  It is now similar to
PrunedLivenessBoundary::visitInsertionPoints and could be used as a
template for a PrunedLivenessBoundary::visitBoundaryPoints through which
::visitInsertionPoints might be factored.

nate-chandler committed:
[CanonicalizeOSSALifetime] Separated steps.

Rather than having finding the boundary be a single combined step,
separate finding the original boundary from extending that boundary.
This enables inserting an optional step between those steps, namely to
extend unconsumed liveness to its original extent at Onone.

nate-chandler committed:
[Gardening] Clarified comment.

nate-chandler committed:
[Gardening] Tweaked whitespace in comment.

nate-chandler committed:
[CanonicalizeOSSALifetime] Extend Onone lifetimes.

To improve the debugging experience of values whose lifetimes are
canonicalized without compromising the semantics expressed in the source
language, when canonicalizing OSSA lifetimes at Onone, lengthen
lifetimes as much as possible without incurring copies that would be
eliminated at O.

rdar://99618502

nate-chandler commented:
@swift-ci please test and merge
